core-bundle/src/Resources/contao/drivers/DC_Table.php

```diff
@@ -6238,6 +6215,100 @@ public function initPicker(PickerInterface $picker)
 
 		return $attributes;
 	}
+
+	/*
+	 * Check whether the given record has a parent that needs to be restored before.
+	 */
+	private function getParentRecordToRestore(string $table, array $data): ?Result
+	{
+		$this->loadDataContainer($table);
+
+		$parent = array();
+
+		if (isset($GLOBALS['TL_DCA'][$table]['config']['dynamicPtable']) && true === $GLOBALS['TL_DCA'][$table]['config']['dynamicPtable'])
+		{
+			$parent = array('table' => $data['ptable'], 'id' => $data['pid']);
+		}
+
+		if (!$parent && isset($GLOBALS['TL_DCA'][$table]['config']['ptable']))
+		{
+			$parent = array('table' => $GLOBALS['TL_DCA'][$table]['config']['ptable'], 'id' => $data['pid']);
+		}
+
+		if (!$parent)
+		{
+			return null;
+		}
+
+		$objParent = $this->Database
+			->prepare('SELECT COUNT(*) AS rowExists FROM ' . $parent['table'] . ' WHERE id = ?')
+			->execute($parent['id'])
+		;
+
+		if ((int) $objParent->rowExists > 0)
+		{
+			return null;
+		}
+
+		$objParentRecord = $this->Database
+			->prepare('SELECT * FROM tl_undo WHERE fromTable = ? AND originalId = ?')
+			->limit(1)
+			->execute($parent['table'], $parent['id'])
+		;
+
+		return ($objParentRecord->numRows > 0) ? $objParentRecord : null;
+	}
+
+	private function restoreUndoRecord(array $data, bool &$log)
+	{
+		$arrFields = array();
+
+		// Restore the data
+		foreach ($data as $table=>$fields)
+		{
+			$this->loadDataContainer($table);
+
+			// Get the currently available fields
+			if (!isset($arrFields[$table]))
+			{
+				$arrFields[$table] = array_flip($this->Database->getFieldNames($table));
+			}
+
+			foreach ($fields as $row)
+			{
+				// Unset fields that no longer exist in the database
+				$row = array_intersect_key($row, $arrFields[$table]);
+
+				// Re-insert the data
+				$objInsertStmt = $this->Database->prepare("INSERT INTO " . $table . " %s")
+					->set($row)
+					->execute();
+
+				// Do not delete record from tl_undo if there is an error
+				if ($objInsertStmt->affectedRows < 1)
+				{
+					$error = true;
+				}
+
+				// Trigger the undo_callback
+				if (\is_array($GLOBALS['TL_DCA'][$table]['config']['onundo_callback']))
+				{
+					foreach ($GLOBALS['TL_DCA'][$table]['config']['onundo_callback'] as $callback)
+					{
+						if (\is_array($callback))
+						{
+							$this->import($callback[0]);
+							$this->{$callback[0]}->{$callback[1]}($table, $row, $this);
```

Comment:
This could be a problem: `$this` is always the DataContainer of the original undo record, that the user intended to restore. But `$table` or `$row` might be a parent record or a parents parent record. Not sure if I should create a new instance of `DataContainer` here?